### ISSUE TITLE
ReferenceError Can't find variable: IntersectionObserver 

### ISSUE THREAD

**Issue by tofra** [tjoskar/ng-lazyload-image#479]
I'm getting some bugreports from Bugsnag using ng-lazyload-image version 9.0.0 with Angular 9.1.12. Any idea why or when this happens? I haven't seen it myself yet.

```
ReferenceError Can't find variable: IntersectionObserver 
    node_modules/ng-lazyload-image/__ivy_ngcc__/fesm2015/ng-lazyload-image.js:312:49 getObservable
    node_modules/ng-lazyload-image/__ivy_ngcc__/fesm2015/ng-lazyload-image.js:128:48 
    node_modules/rxjs/_esm2015/internal/operators/switchMap.js:28:34 _next
    node_modules/rxjs/_esm2015/internal/Subscriber.js:49:23 next
    node_modules/rxjs/_esm2015/internal/operators/tap.js:46:30 _next
    node_modules/rxjs/_esm2015/internal/Subscriber.js:49:23 next
    node_modules/rxjs/_esm2015/internal/operators/tap.js:46:30 _next
    node_modules/rxjs/_esm2015/internal/Subscriber.js:49:23 next
    node_modules/rxjs/_esm2015/internal/operators/tap.js:46:30 _next
    node_modules/rxjs/_esm2015/internal/Subscriber.js:49:23 next
    node_modules/rxjs/_esm2015/internal/ReplaySubject.js:57:32 _subscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:42:35 _trySubscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:28:35 subscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:23:35 subscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:23:35 subscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:23:35 subscribe
    node_modules/rxjs/_esm2015/internal/Observable.js:23:35 subscribe
    node_modules/ng-lazyload-image/__ivy_ngcc__/fesm2015/ng-lazyload-image.js:130:27 
    node_modules/zone.js/dist/zone-evergreen.js:123:53 run
    node_modules/ng-lazyload-image/__ivy_ngcc__/fesm2015/ng-lazyload-image.js:122:38 ngAfterContentInit
    node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:4726:22 callHook
...
```

**Comment by tofra** [comment]
It happens only on iOS and I see it doesn't happen on iOS >= 13.0
![image](https://user-images.githubusercontent.com/1746173/93793778-91ef4780-fc37-11ea-9d75-86afb091c037.png)

**Comment by tjoskar** [comment]
Hi @tofra,

You need to include a polyfill for iOS12.1 and below: https://caniuse.com/intersectionobserver

You can find a polyfill here: https://github.com/w3c/IntersectionObserver/tree/master/polyfill or using [https://polyfill.io](polyfill.io) to only include it in needed browsers: https://polyfill.io/v3/polyfill.min.js?features=IntersectionObserver%2CIntersectionObserverEntry

**Comment by tjoskar** [comment]
> The polyfill.io generates an empty file

It will return an empty file if you download the file in a browser that supports intersection-observer. So it will not be empty in safari for iOS12.1 or below. 

> I do see that onStateChange event is called only twice

What browser are you using when you see that? Do you get that behavior in Chrome/Firefox on desktop, with and without your polyfill?

**Comment by tofra** [comment]
> > The polyfill.io generates an empty file
> >
> It will return an empty file if you download the file in a browser that supports intersection-observer. So it will not be empty in safari on iOS12.1 or below.

Ah! Correct, when I load it in the emulator with iOS11 it indeed contains the script.

> > I do see that onStateChange event is called only twice
> 
> What browser are you using when you see that? Do you get that behavior in Chrome/Firefox on desktop, with and without your polyfill?

I'm running Safari in an Emulator with iOS11.

I have tried to make 3 example projects trying to figure out where the issue could be. I made 2 examples on stackblitz (I always use this to "quickly" check things to eliminate that the problem is in my project), and I made 1 example in github creating the ionic blank starter template with angular 10. The two examples on stackblitz work, the one in github doesn't.

Do you have an idea why it doesn't work in the example github (https://github.com/tofra/ng-lazyload-image-test) which is just a blank Angular/Ionic project?
Let me know if you need more examples!

Kind regards,
Tom

---

Angular project works on Stackblitz (Angular 10)
https://angular-ivy-9ydyxo.stackblitz.io
https://stackblitz.com/edit/angular-ivy-9ydyxo
![image](https://user-images.githubusercontent.com/1746173/94135220-69e42c00-fe63-11ea-896b-6d21b87d5bc0.png)

Ionic project works on Stackblitz (Angular 5, Ionic 3.9.2)
https://ionic-zxzjnk.stackblitz.io
https://stackblitz.com/edit/ionic-zxzjnk
![image](https://user-images.githubusercontent.com/1746173/94135270-87b19100-fe63-11ea-9c4b-6a743e16ab27.png)

Ionic project doesn't work (Angular 10, Ionic 5) 
Github: https://github.com/tofra/ng-lazyload-image-test
![image](https://user-images.githubusercontent.com/1746173/94135315-96984380-fe63-11ea-9a6c-4be3c8b2b2d0.png)

In Safari on my Mac it works (obviously as `intersection-observer` is known in Safari 14.0)
![image](https://user-images.githubusercontent.com/1746173/94135356-a57ef600-fe63-11ea-80b8-c6f2f38460ec.png)

**Comment by tjoskar** [comment]
Thanks for your detailed report. I will take a look and see if I can find anything.

**Comment by tofra** [comment]
@tjoskar where you able to find something?

**Comment by tjoskar** [comment]
@tofra, sorry for the late reply. I took a look this morning and this is what I found:

First of, if we open up `node_modules/intersection-observer/intersection-observer.js` and comment out the following lines:
```
if ('IntersectionObserver' in window &&
    'IntersectionObserverEntry' in window &&
    'intersectionRatio' in window.IntersectionObserverEntry.prototype) {

  // Minimal polyfill for Edge 15's lack of `isIntersecting`
  // See: https://github.com/w3c/IntersectionObserver/issues/211
  if (!('isIntersecting' in window.IntersectionObserverEntry.prototype)) {
    Object.defineProperty(window.IntersectionObserverEntry.prototype,
      'isIntersecting', {
      get: function () {
        return this.intersectionRatio > 0;
      }
    });
  }
  return;
}
```
The polyfill will kick in regardless of the browser support `IntersectionObserver` or not. This makes it easier to debug in Chrome/Firefox.

I then started to debug the flow and I can't really understand what the issue is but it is something to do with ShadowDom and the fact that ionic is using `position: absolute` that creates parents that have size 0 (height and weight). 

I tried to add a naive fix in `getParentNode` for `intersection-observer.js`: 
```ts
if (node.host) {
  return node.host;
}
```

And that makes it to work as long as the parent node is not a `position: absolute`.

So I really don't know what the issue is but it has something to do with the combination of Ionic and the polyfill.

https://github.com/w3c/IntersectionObserver/issues/424
